### DISCUSSION TITLE
请问哪里有教程或者文档指导我们基于AntD进行扩展开发组件呢？

### DISCUSSION THREAD

**Discussion by liaodalin19903** [ant-design/ant-design#41589] [Q&A]
请问哪里有教程或者文档指导我们基于AntD进行扩展开发组件呢？
1.想要TS代码写
2.想要函数式组件

**Comment by Wxh16144** [comment]
仅代表个人观点，这是我之前基于 antd@4 搭建的 monorepo 脚手架，你可以参考一下 [react-ui-library](https://github.com/template-pro/react-ui-library)、[template-pro/react-ui-library 模板使用示例_哔哩哔哩_bilibili](https://www.bilibili.com/video/BV1ND4y1L7Nf)。不适用于 antd@5, 后续我会更新一版~

**Comment by vaynevayne** [comment]
@liaodalin19903, 有兴趣可以看源码

毛遂自荐  https://www.npmjs.com/package/@soul/core  目前功能如下
扩展 table, 支持行列拖拽排序, 列显隐, excel导出
扩展 select, 支持添加预设,反选 模糊选

二次封装很简单
关键点就几个:
1. 透传antd属性 ...tableProps
2. 扩展属性放统一命名空间下, 比如你组件库名字 soul 
3. 处理默认值;  注意引用类型, 不要使用函数参数默认值或对象解构赋值默认值,因为会漏出来null
4. 支持受控和非受控两种使用方式, 使用 [useUncontrolled](https://mantine.dev/hooks/use-uncontrolled/)
5. 如无必要, 进出的数据结构尽量保持一致, 比如传入 `{}` 传出 `{name:''}` 是错误的, 会导致原属性被覆盖
6. (可选)数据转化, 比如为了方便跟后端交互,使用字符串并且包含多条信息, `'excludes:age,name'`,  组件在内部可以拆成数组使用,下面的例子age就是数组, onChange时再合为一个字符串,parseValue,transformValue
7. (可选)派生状态, 比如封装table,要使用到 ` table.columns`, 内部希望可以添加 `order` 属性到这上面, 而且内部受控, prop改变时再使用prop覆盖默认值, 这在vue很常见,也很简单  react 中,因为引用变了,props就更新,内部状态就没了, 所以你可以使用  [useWatch](https://www.npmjs.com/package/@soul/utils) + useMemo/ compareFn(),  自己使用, 直接compareFn也行, antd pro源码也是这样的,  
不过 @sour/core 中为了支持可回显, 我选择了 useWatch + 另一个State =  合并成新的state,达到同样的效果

为了方便使用, 我举一个稍微复杂点的例子


Form.tsx
```jsx
 // 这是非受控的用法, 也支持受控
<MyInput  soul={{
   age:'excludes:age,name',
   onAgeChange:  console.log
}}  />
```

MyInput.tsx
```jsx
const parseValue = (value)=> value.spilt(':') //  'excludes:age' => ['excludes', 'age,name' ] 此处切割为数组, 你切成对象也是可以的
const transformValue = (val)=> String(val) //  ['excludes', 'age,name' ] => 'includes:age'

type MyInputProps = {
 soul: {
   age:string,
   onAgeChange:(age:string)=> void
 }
} & inputProps

const MyInput  = ({soul, ...inputProps}:MyInputProps)=>{

const [spiltedAge, setSplitedAge] = useUncontrolled({
  value:  parseValue(soul.age) ,
  onChange: (val)=> soul.onAgeChange(transformValue(val)),
  defaultValue: []
})

// 读
console.log( spiltedAge[0])
// 写
const onModeChange=(newMode)=> setSplitedAge([newMode, spiltedAge[1]])

return (
<>
  <Button onClick={()=> onModeChange('includes')}>includes</Button>
  <Input {...inputProps} />
</>
)
}
```

希望我表达清楚了
